### ISSUE TITLE
max listener limit hit with live replication

### ISSUE THREAD

**Issue by prust** [pouchdb/pouchdb#3848]
I've been following PouchDB for some time but I've only recently started using it so there may be a boneheaded mistake in my code. I'm using one memdown PouchDB on the server and one websql PouchDB in Chrome, doing live 2-way replication.

I'm running into two problems that are probably related. The first is that I get unexpected "pause" events in the browser, which makes me nervous (I'm running on localhost, so there shouldn't be a connection problem), and another is that I hit the `possible EventEmitter memory leak detected. 11 listeners added` error on the server if I hit refresh 11x with an empty DB.

Here is `server.js`:

```javascript
var express = require('express');
var app = express();
var PouchDB = require('pouchdb');
var expressPouch = require('express-pouchdb');
var memdown = require('memdown');

var InMemPouchDB = PouchDB.defaults({db: memdown});
 
app.use('/db', expressPouch(InMemPouchDB, {'mode': 'minimumForPouchDB'}));

app.use(express.static('./'));

var server = app.listen(8083, function() {
  var host = server.address().address;
  var port = server.address().port;
  console.log('Example app listening at http://%s:%s', host, port);
});

new InMemPouchDB('db_csnw');
```

Here is `index.html`:

```html
<!DOCTYPE html>
<script src="pouchdb.js"></script>
<script>
document.addEventListener("DOMContentLoaded", function() {
  var local_db = new PouchDB("db_csnw", {adapter: "websql"});
  var remote_db = new PouchDB("http://localhost:8083/db/db_csnw");
  var sync = local_db.sync(remote_db, {
    live: true
  }).on("change", function (change) {
    console.log("something changed");
  }).on("paused", function (info) {
    console.log("replication was paused, usually because of a lost connection");
  }).on("active", function (info) {
    console.log("replication was resumed");
  }).on("error", function (err) {
    console.error(err);
  });
  window.addEventListener("beforeunload", function() {
    sync.cancel();
  });
});
</script>
```

If I run this with node w/ the latest versions of express (`4.12.3`), express-pouch (`0.14.0`), memdown (`1.0.0`) and pouch (`master`).

I thought that maybe #3832 or #3825 would fix this, so I upgraded client and server from `3.5.0` to `master`, but I'm still getting the same behavior (though I think I'm running into slightly fewer "pause" events).

Any idea what I can do to debug this further? (if it helps, I'm running Win7, Chrome 42 and Node 0.10.32)

**Comment by nolanlawson** [comment]
> if I hit refresh 11x with an empty DB.

So this actually sounds related to https://github.com/pouchdb/pouchdb/issues/3847. Maybe we should indeed be canceling replication during the `beforeunload` event.

Although you yourself are canceling sync in the `beforeunload`, so I'm not sure which event emitter is getting so many listeners attached to it. Does the error output tell you?

**Comment by nolanlawson** [comment]
Ah wait, I'll bet the issue is that `cancel()` doesn't remove the listeners synchronously. The `beforeunload` event is supposed to do all its work synchronously, so if anything is async, then the listeners won't get removed.

**Comment by nolanlawson** [comment]
As for `paused` and `active` firing too many times: it's something I'm aware of and working on right now.

**Comment by nolanlawson** [comment]
If you don't mind, I think this issue can be closed in favor of #3852 and https://github.com/pouchdb/pouchdb/issues/3847. In particular, I think fixing https://github.com/pouchdb/pouchdb/issues/3847 would solve the max listener problem. Let me know if I missed anything from this issue.

**Comment by nolanlawson** [comment]
Actually wait, this is different from https://github.com/pouchdb/pouchdb/issues/3847. The error thrown in that case is a 500.

**Comment by prust** [comment]
@nolanlawson:

Thanks for looking into this!

> As for paused and active firing too many times: it's something I'm aware of and working on right now

Cool, good to know that I'm not the only one seeing that.

> I'm not sure which event emitter is getting so many listeners attached to it. Does the error output tell you?

There is a stack trace, I should've thought to include it before. Here it is (I stripped out my current project folder from the stack trace, so it starts at pouchdb):

```
at Changes.addListener [as on] (events.js:160:15)
at Changes.addListener (pouchdb\lib\utils.js:224:8)
at PouchAlt.LevelPouch.api._changes (pouchdb\lib\adapters\leveldb\leveldb.js:882:26)
at Changes.doChanges (pouchdb\lib\changes.js:181:28)
at new Changes (pouchdb\lib\changes.js:94:10)
at PouchAlt.AbstractPouchDB.changes (pouchdb\lib\adapter.js:629:10)
at Changes.<anonymous> (express-pouchdb\lib\routes\changes.js:67:34)
at Changes.emit (events.js:117:20)
at pouchdb\lib\changes.js:21:12
at pouchdb\lib\utils.js:329:11
at pouchdb\node_modules\argsarray\index.js:14:18
at pouchdb\lib\changes.js:80:5
at tryCatch1 (pouchdb\node_modules\bluebird\js\main\util.js:63:19)
at Promise$_callHandler [as _callHandler] (pouchdb\node_modules\bluebird\js\main\promise.js:695:13)
at Promise$_settlePromiseFromHandler [as _settlePromiseFromHandler] (pouchdb\node_modules\bluebird\js\main\promise.js:711:18)
at Promise$_settlePromiseAt [as _settlePromiseAt] (pouchdb\node_modules\bluebird\js\main\promise.js:868:14)
```

I stuck a console.log() inside leveldb.js's `api._changes()` method, where it calls `LevelPouch.Changes.addListener(...)`, so I could see the name and id that it is passing. The name is always the name of my database (`db_csnw`) and the id starts with `"db_csnw:"`, followed by a different uuid each time.

> The error thrown in that case is a 500

I also get a 500 error if, instead of refreshing the browser, I just sit and wait for 30 seconds. It seems to be triggered by this long poll getting `(canceled)` after 30sec:

![image](https://cloud.githubusercontent.com/assets/23037/7671036/e03d9eae-fc70-11e4-9905-69e753851ee9.png)

If I add `retry: true` then instead of a 500, I get a "pause" event on the client every 30 seconds and (if I leave the server-side console.log() in leveldb.js mentioned above, I can see addListener() getting called every 30 seconds with each _changes long-poll retry) then after 10x (300sec) it hits the same EventEmitter listener limit warning with the same callstack as above.

**Comment by nolanlawson** [comment]
Sorry, I tried with both PouchDB Server and CouchDB, but I can't reproduce your issue no matter how many times I refresh (using Cmd-R on my Mac in Chromium).

Here is the live example I used to try to reproduce: http://bl.ocks.org/nolanlawson/e6ce113e4cfa8740f7ad

**Comment by prust** [comment]
@nolanlawson: It's hard to tell for sure, since this error is thrown on the server and bl.ocks only exposes the client, but I think the reason you're not getting the error is related to how your server returns a `200` success response for `_changes`:

![image](https://cloud.githubusercontent.com/assets/23037/7690026/62751cfe-fd65-11e4-9127-99a96effe94c.png)

But my server is returning a `(canceled)` error (maybe the request is canceled client-side?):

![image](https://cloud.githubusercontent.com/assets/23037/7690078/d6d7a530-fd65-11e4-915f-3414cb97adfb.png)

Notice also that mine is exactly `30.00` seconds while yours is `25.24`, maybe mine is being canceled by the client after a 30sec timeout but yours is being finished by the server after 25 seconds?

* * *

I put the code that reproduces the issue on github (https://github.com/prust/pouchdb-issue-3848) and ran it on a mac via:

```
git clone https://github.com/prust/pouchdb-issue-3848.git
cd pouchdb-issue-3848
npm install
node server.js
```

I got the same error on the mac (it was thrown on the server on the 11th refresh).

If I add `retry: true` to the client then I get the server-side error without needing to refresh at all, I just need to wait 5 minutes for the 11th `(canceled)` _changes.

**Comment by nolanlawson** [comment]
This should be fixed in 3.6.0. Please reopen if it's not.

**Comment by prust** [comment]
Thanks @nolanlawson!

I'm not in a good position to double-check at the moment, but I'll take a look at some point.

Thanks again for your work on PouchDB and your blog posts about Pouch and WebSQL/IndexedDB, it really is helpful.
